wc プログラムを作成

shizuka-yamamoto opened the pull request.

shizuka-yamamoto committed:
wcプログラムの作成

shizuka-yamamoto committed:
shebangの追加

shizuka-yamamoto committed:
rubocopの実行&コード修正

u1tnk commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,51 @@
+def main
+  l_option = ARGV.getopts('l')
+  if ARGV[0]
+    argument_input(l_option)
```
名前の話ですが、`xxx_input` という名前だと「入力を処理するのかな？」となってしまうので、内容が類推できる名前にしましょう

u1tnk reviewed (changes requested):
メソッドで、重複処理を整理するあたりは申し分無いですね。

コメントした名前づけのところを修正すれば十分ですが、

全体の流れがメイン処理で分かるともっと良いですね

インプットの処理
↓
データ収集
↓
出力

分け方にもいくつかありますが、おおまかにこんな流れになっていて、プログラムを読む方が理解しやすいようになるととても良くなりますね。

shizuka-yamamoto committed:
メソッド名の修正

shizuka-yamamoto commented:
@u1tnk 
コードレビューをありがとうございます！
メソッド名を修正しましたので、ご確認をよろしくお願いします。

また、アドバイスいただいた、`main`メソッドに全体の流れを集約する件ですが、
出力処理を`main`メソッドに持ってくるという点が特に難しく行き詰まっており、現状の方向性がよいものか、ご相談をさせてください🙇‍♀️

■確認したいこと
* 引数にファイル名が1つの場合と2以上の場合で出力の処理に違いがあるため、`main`メソッドで条件分岐する
* 現状のコードの方向性の確認
標準入力の場合とファイル名が1つの場合のみ`main`メソッドへインプット〜出力までの処理を書いてみました。
`main`メソッドに出力の処理をもってくるという考え方はあってますでしょうか？🙇‍♀️

```ruby
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'optparse'

def main
  l_option = ARGV.getopts('l')
  if ARGV[0]  # ファイル1つ指定の時の処理
    text = File.read(ARGV[0])  # インプット処理
    calc_of_line_word_byte(text)  # データの収集
    return printf("%8d #{ARGV[0]}\n", @line) if l_option['l']  # データの出力
    printf("%8d%8d%8d #{ARGV[0]}\n", @line, @word, @byte)
  elsif ARGV[1] # ファイル複数指定ある場合の処理

  else  # 標準入力
    standard_input  # インプット処理
    calc_of_line_word_byte(@text)  #データの収集
    return puts @line.to_s.rjust(8) if l_option['l']  # データの出力
    printf("%8d%8d%8d\n", @line, @word, @byte)
  end
end


# def file_specification(l_option)
#   # total_count_lines = total_count_words = total_count_bytes = 0

#   ARGV.each do |file_name|
#     text = File.read(file_name)
#     # total_count_lines += text.count("\n")
#     # total_count_words += text.split(/\s+/).length
#     # total_count_bytes += text.bytesize
#     calc_of_line_word_byte(text)
#     puts " #{file_name}"
#   end

#   # return unless ARGV.size > 1 # ファイルが複数ある場合は合計を出力

#   # print total_count_lines.to_s.rjust(8)
#   # unless l_option['l']
#   #   print total_count_words.to_s.rjust(8)
#   #   print total_count_bytes.to_s.rjust(8)
#   # end
#   # puts ' total'
# end

def standard_input
  @text = $stdin.read
end

def calc_of_line_word_byte(text)
  @line = text.count("\n")
  @word = text.split(/\s+/).length
  @byte = text.bytesize
end

main

```

u1tnk commented:
ちょっと現状だとメイン処理にまとめはしてますが全体がifで別れているので、流れが分かりづらくなっていますね。

こういうときの為にブログ書いてたのを忘れてました。
https://u1tnk.github.io/blog/2020/12/17/structured_programming/

参考になると思うので見てみてください。

入力の処理のときに、ファイル/標準入力に関わらず、一回変数に保存しておいて、その後を共通化するって感じですね。
出力のときにも分岐する必要がありますが、分岐の為の情報は保存しておいた変数にあるので、そこだけ見れば良いはずですよ 💪

shizuka-yamamoto committed:
mainメソッドで全体のプログラムが把握できるように修正する

shizuka-yamamoto commented:
@u1tnk 
`main`メソッドで全体が把握できるように、修正をしました。
ご確認をよろしくお願い致します🙇‍♀️
このコードを作るのに、結構苦労してしまいました😣

■苦労した点とどう対応したか
* 入力処理（`parse_input`メソッド内）で、ファイル指定があった際、どこまでの情報を変数として保存しておくべきか
  * 今回は、「ファイル名」のみを取得することで、データ収集の際に複数のファイル指定あった場合に対応できるようにする
* データ収集時、ファイル指定のあるデータをどのように変数として保存しておくか
  * ファイル名を含めた配列として保存することで対応する

u1tnk commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,67 @@
+
+def parse_input
+  if ARGV.size >= 1 # ファイル指定あり
+    @files = ARGV.map do |file_name|
```
クラスを作ってないときにインスタンス変数を使うのはあまり推奨できません。
この場合はparse_inputの結果をまとめて返して、mainメソッド内の変数でやりとりした方が分かりやすいと思います。

本来色々やり方はありますが、ちょっと苦戦されてるようなのでアドバイスすると、標準入力とファイル入力と同じレベルに合わせて差異を無くすと後で処理しやすいので、標準入力に合わせてファイルの中身も読んでしまって良いと思います。

で、まとめて返せば良いです。
今の `@files`/`@text`の情報に加えてファイルの中身を読みこんだものを返せば良いです。
複数変数returnしても良いですが、一つのhashにまとめるのも扱いやすくて良いと思います。

以下のような配列だと欲しい情報は入っていると思います。標準入力のときは file_nameは無いですが、表示もしないので問題ありません。
```
[
  {
     file_name: "",
     text: ....
  },
]
```

また、calcとparse_inputは分けても良いですが、処理がシンプルになるなら合わせても良いですよ。
合格後に見られる私が作ったお手本はそこは一緒になっています。参考まで。

u1tnk commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,67 @@
+end
+
+def calc
+  if ARGV.size >= 1 # ファイル指定あり
```
parse_inputの方で指摘した通り入力に関する処理は parse_inputのみにしないと分けた意味が無いのでここでARGVを参照するのは良くないですね。
parse_inputでデータ読み込みが終わっていれば、標準入力のときと指定したファイルが一つのときはこれから後は共通化できると思いますよ 💪

u1tnk reviewed (changes requested):
コメントしましたのでご確認ください。

shizuka-yamamoto committed:
標準入力のレベルに合わせて変数textにデータを保存する

shizuka-yamamoto committed:
calcメソッド内の計算処理の修正

shizuka-yamamoto committed:
parse_inputの内容を変数numbersに保存する

shizuka-yamamoto committed:
formatメソッド内を修正

shizuka-yamamoto committed:
rubocopの実行＆修正

shizuka-yamamoto commented:
@u1tnk 
コードレビューをありがとうございました！！
そして、再考しましたので、ご確認をよろしくお願い致します🙇‍♀️
アドバイスいただいた通り、hashを使いデータを整理することで、大分読みやすくなったかなと思います😭

■確認したいこと
* `main`メソッド内で、`parse_input`をそのまま他の変数に代入していますがこのような書き方はありなのでしょうか・・？

u1tnk commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,62 @@
+
+def main
+  l_option = ARGV.getopts('l')
+  numbers = parse_input
```
メソッド名ですが、 inputをparseしたら numbersができたとなって読み手が混乱してしまいます。
後述しますが、calcは外の方が良さそうですね。

u1tnk commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,62 @@
+end
+
+def parse_input
+  text = if ARGV.size >= 1 # ファイル指定あり
```
ここで`text = ` を外して calcの呼び出しを消せば、parseした結果をそのまま返せます。
返ってきた値を `inputs `とかの変数名で一回受けて calcに渡してnumbersが返ってきた方が素直だと思います。

u1tnk reviewed (changes requested):
少しだけ構成にコメントしましたが、コードの重複もほぼなくなり、スッキリしましたね 👍

shizuka-yamamoto committed:
parse_inputの結果を変数inputsに保存するよう修正

shizuka-yamamoto committed:
calcメソッドの戻り値を変数numbersに保存するよう修正

shizuka-yamamoto commented:
@u1tnk 
コードレビューをありがとうございます！！
修正しておりますので、ご確認をよろしくお願い致します🙇‍♀️
変数名からどんな操作を行っているか、想像できるようなコードを追求していきます🙇‍♀️💪

u1tnk reviewed (approved):
🙆‍♀️
